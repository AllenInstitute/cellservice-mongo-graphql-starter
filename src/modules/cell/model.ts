import { getModelForClass } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";

import { Cell } from "../../entities";
import { CellInput } from "./input";

// This generates the mongoose model for us
export const CellMongooseModel = getModelForClass(Cell);

export default class CellModel {
  async getById(_id: ObjectId): Promise<Cell | null> {
    // Use mongoose as usual
    return CellMongooseModel.findById(_id).lean().exec();
  }

  async getManyById(_ids: ObjectId[]): Promise<Cell[] | null> {
    return CellMongooseModel.find().where('_id').in(_ids).lean().exec();
  }

  async create(data: CellInput): Promise<Cell> {
    const cell = new CellMongooseModel(data);

    return cell.save();
  }
}
