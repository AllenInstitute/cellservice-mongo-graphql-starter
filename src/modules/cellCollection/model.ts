import { getModelForClass } from "@typegoose/typegoose";
import { isDefined, isNotEmpty } from "class-validator";
import { ObjectId } from "mongodb";

import { CellCollection, Cell } from "../../entities";
import { CellCollectionInput } from "./input";

// This generates the mongoose model for us
export const CellCollectionMongooseModel = getModelForClass(CellCollection);
const CellModel = getModelForClass(Cell)

export default class CellCollectionModel {
  async getById(_id: ObjectId): Promise<CellCollection | null> {
    // Use mongoose as usual
    return CellCollectionMongooseModel.findById(_id).lean().exec();
  }

  async getAll(limit?: number): Promise<CellCollection[] | null> {
    var query = CellCollectionMongooseModel
      .find({})
      .sort({'_id': -1})
      .lean()
    if (limit != null) {
      query = query.limit(limit)
    }
    return query.exec();
  }

  async create(data: CellCollectionInput): Promise<CellCollection> {
    // Throw an error if there's a missing cell
    const missingCells = await Promise.all(data.cells.map(
      async (_id: string): Promise<string | void> => {
        const exists = await CellModel.exists({_id: _id});
        if (exists === false) {
          return _id
        }
      })
    ).then((_) => _.filter(isDefined));
    if (missingCells.length > 0) {
      throw Error(`Tried to create a collection with ${missingCells.length} nonexistent cell _id(s): ${missingCells}`)
    }
    const collection = new CellCollectionMongooseModel(data);
    collection.save();
    // Convert to object, otherwise will get undefined for cells
    return collection.toObject()
  }
}
