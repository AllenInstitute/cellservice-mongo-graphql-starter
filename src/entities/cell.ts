import { ObjectType, Field } from "type-graphql";
import { prop } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";

@ObjectType()
export class Cell {
  @Field()
  readonly _id!: ObjectId;

  @prop()
  @Field(() => Date)
  createdAt!: Date;

  @prop()
  @Field(() => Date)
  updatedAt!: Date;

  @prop()
  @Field()
  type!: string;

}
