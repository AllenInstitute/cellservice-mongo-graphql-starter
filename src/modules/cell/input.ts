import { Field, InputType, ID } from "type-graphql";
import { MaxLength, MinLength } from "class-validator";

@InputType()
export class CellInput {
  @Field()
  @MaxLength(300)
  @MinLength(1)
  type: string;

}
