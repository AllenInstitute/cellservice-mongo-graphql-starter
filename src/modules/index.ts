import CellResolver from "./cell/resolver";
import TodoResolver from "./cell/resolver";
import CollectionResolver from "./cellCollection/resolver";

// Important: Add all your module's resolver in this
export const resolvers: [Function, ...Function[]] = [
  CellResolver,
  CollectionResolver
  // UserResolver
  // AuthResolver
  // ...
];
