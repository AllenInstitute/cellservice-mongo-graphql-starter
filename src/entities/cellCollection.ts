import { ObjectType, Field, ID } from "type-graphql";
import { prop } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";
import { Cell } from './cell'
import { Ref } from '../utils/types'

@ObjectType()
export class CellCollection {
  // @Field(type => ID)
  @Field()
  readonly _id!: ObjectId;

  @prop({default: new Date(), required: true})
  @Field(() => Date)
  createdAt!: Date;

  @prop({default: new Date(), required: true})
  @Field(() => Date)
  updatedAt!: Date;

  @prop({required: true})
  @Field()
  name!: string;

  @prop({ref: Cell, required: true})
  @Field(type => [Cell])
  cells: Ref<Cell>[];

}
