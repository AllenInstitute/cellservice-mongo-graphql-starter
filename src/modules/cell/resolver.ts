import { Resolver, Arg, Query, Mutation, ID } from "type-graphql";
import { Service } from "typedi";
import { ObjectId } from "mongodb";

import { Cell } from "../../entities";
import CellService from "./service";
import { CellInput } from "./input";

/*
  IMPORTANT: Your business logic must be in the service!
*/

@Service() // Dependencies injection
@Resolver((of) => Cell)
export default class CellResolver {
  constructor(private readonly todoService: CellService) {}

  @Query((returns) => Cell)
  async getCell(@Arg("id") id: ObjectId) {
    const todo = await this.todoService.getById(id);

    return todo;
  }

  @Mutation((returns) => Cell)
  async createCell(
    @Arg("cellData") cellData: CellInput
  ): Promise<Cell> {
    const todo = await this.todoService.addCell(cellData);
    return todo;
  }
}
