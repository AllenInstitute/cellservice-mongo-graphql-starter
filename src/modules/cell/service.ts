import { Service } from "typedi";
import { ObjectId } from "mongodb";

import CellModel from "./model";
import { Cell } from "../../entities";
import { CellInput } from "./input";

@Service() // Dependencies injection
export default class CellService {
  constructor(private readonly cellModel: CellModel) {}

  public async getById(_id: ObjectId): Promise<Cell | null> {
    return this.cellModel.getById(_id);
  }

  public async addCell(data: CellInput): Promise<Cell> {
    const newCell = await this.cellModel.create(data);

    // Business logic goes here
    // Example:
    // Trigger push notification, analytics, ...

    return newCell;
  }
}
