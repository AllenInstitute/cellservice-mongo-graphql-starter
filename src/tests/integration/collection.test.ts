import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server-express";
import { buildSchema } from "../../utils";
import mongoose from "mongoose";

import { resolvers } from "../../modules";
import { CellMongooseModel } from "../../modules/cell/model";
import { CellCollectionMongooseModel } from "../../modules/cellCollection/model";


import {
  connect,
  clearDatabase,
  closeDatabase,
  populateDatabase,
} from "../utils";

// Set the cell Ids ahead of time for reference in other tests
const cellIds = [
  new mongoose.Types.ObjectId().toHexString().toString(),
  new mongoose.Types.ObjectId().toHexString().toString()
]

beforeAll(async () => connect());

// You can populate the DB before each test
beforeEach(async () => {

  await populateDatabase(CellMongooseModel, [
    {
      _id: cellIds[0],
      type: "t-type",
    },
    {
      _id: cellIds[1],
      type: "d-type",
    },
  ]);
});

afterEach(async () => {
  await clearDatabase();
});

afterAll(async (done) => {
  await closeDatabase();
  done();
});

/**
 * Prompt test suite.
 */
describe("Collection", () => {
  it(`should create a cell collection`, async () => {
    // We build the schema
    const graphQLSchema = await buildSchema();

    // We create the Apollo Server
    const server = new ApolloServer({
      schema: graphQLSchema,
    }) as any;

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    // We define the query and the variables as you would do from your front-end
    const variables = {
      collectionData: {
        name: "test collection",
        cells: cellIds,
      },
    };

    const CREATE_COLLECTION = gql`
      mutation createCollection($collectionData: CellCollectionInput!) {
        createCollection(collectionData: $collectionData) {
          name
          cells {
            type
          }
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await mutate({
      mutation: CREATE_COLLECTION,
      variables,
    });

    expect(res).toMatchSnapshot();
  });

  it(`should get a cell collection`, async () => {
    // We generate a collection ID
    const collectionId = new mongoose.Types.ObjectId().toHexString().toString();

    // Add a cell collection with the generated ID in the database
    await populateDatabase(CellCollectionMongooseModel, [
      {
        _id: collectionId,
        name: "another collection",
        cells: cellIds
      },
    ]);

    // We build the schema
    const graphQLSchema = await buildSchema();

    // We create the Apollo Server
    const server = new ApolloServer({
      schema: graphQLSchema,
    }) as any;

    // use the test server to create a query function
    const { query } = createTestClient(server);

    const variables = {
      id: collectionId,
    };

    const GET_CELL = gql`
      query getCollection($id: ObjectId!) {
        getCollection(id: $id) {
          name
          cells {
            type
          }
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await query({
      query: GET_CELL,
      variables,
    });

    expect(res).toMatchSnapshot();
  });
});


it(`should fail if a cell doesn't exist`, async () => {

    // We generate a fake cell ID
    const fakeId = new mongoose.Types.ObjectId().toHexString().toString();
    const cellIdsWithFake = cellIds.concat(fakeId)
    // We build the schema
    const graphQLSchema = await buildSchema();

    // We create the Apollo Server
    const server = new ApolloServer({
      schema: graphQLSchema,
    }) as any;

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    // We define the query and the variables as you would do from your front-end
    const variables = {
      collectionData: {
        name: "test collection",
        cells: cellIdsWithFake,
      },
    };

    const CREATE_COLLECTION = gql`
      mutation createCollection($collectionData: CellCollectionInput!) {
        createCollection(collectionData: $collectionData) {
          name
          cells {
            type
          }
        }
      }
    `;

    // run query against the server
    const res = await mutate({
      mutation: CREATE_COLLECTION,
      variables,
    });

    const expectedMessage = `Tried to create a collection with 1 nonexistent cell _id(s): ${fakeId}`
    expect(res.data).toBeNull();
    expect(res.errors).not.toBeNull();
    expect(res.errors!.length).toEqual(1);
    expect(res.errors![0].message).toEqual(expectedMessage)
});
