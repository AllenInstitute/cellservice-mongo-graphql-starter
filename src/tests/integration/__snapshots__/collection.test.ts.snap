// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection should create a cell collection 1`] = `
Object {
  "data": Object {
    "createCollection": Object {
      "cells": Array [
        Object {
          "type": "t-type",
        },
        Object {
          "type": "d-type",
        },
      ],
      "name": "test collection",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Collection should get a cell collection 1`] = `
Object {
  "data": Object {
    "getCollection": Object {
      "cells": Array [
        Object {
          "type": "t-type",
        },
        Object {
          "type": "d-type",
        },
      ],
      "name": "another collection",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
